Add Defense articles (#4)
xsleaks/xsleaks.github.io-sources: content/docs/defenses/_index.md

title: "Defense Mechanisms"
---

# Defense Mechanisms

Defending against all [XS-Leaks Attack Vectors]({{< ref "../attacks/" >}}) is not a trivial task. Each one of them affects different web and browser components and has its quirks. Some Bug Bounty programs, such as Google VRP even stopped receiving and paying for new XS-Leaks reports since they believe it's not feasible to prepare a [particular and unique bugfix]({{< relref "_index.md#application-design" >}}) for each report [^1]. Instead, Google and other companies believe the right approach towards fixing XS-Leaks is to invest time and human power into proposing [new large scale mitigations]({{< relref "_index.md#opt-in-mechanisms" >}}), so that applications can use them to and fix considerable chunks of XS-Leaks.

## Application Design

Application Design [techniques]({{< ref "design-protections/_index.md" >}}) allow developers to better use certain `HTML APIs` known to easily introduce XS-Leaks. This type of techniques help develop temporary measures when [opt in mechanisms]({{< relref "_index.md#opt-in-mechanisms" >}}) can't be deployed in the application right away.

{{< hint warning >}}

> These [techniques]({{< ref "design-protections/_index.md" >}}) should be seen as **temporary solutions**, before  [opt in mechanisms]({{< relref "_index.md#opt-in-mechanisms" >}}) are deployed. Only useful in particular situations, and should not be the default strategy.

Author: e0925f0 fixed this

{{< /hint >}}

## Browser Intrinsic

> Browsers actively discuss how to incorporate [new mitigations]({{< relref "_index.md#browser-intrinsic" >}}) transparent to both users and developers, that help mitigate some of the XS-Leaks mentioned in this Wiki.

Reviewer: ```suggestion
Browsers vendors actively discuss how to implement [new mechanisms]({{< relref "_index.md#browser-intrinsic" >}}) to help mitigate some of the XS-Leaks mentioned in this Wiki. They always try to make such mechanisms as transparent as possible to users and developers but this is not always possible.
```

As much as I would like to see [a discussion between browsers](https://www.webtoons.com/en/challenge/internet-explorer/list?title_no=219164&page=1), I don't think that is what you meant.

{{< hint info >}}

> Can be highly effective. The biggest advantage is that it transparently protects applications and users, and should not require extra efforts from developers.

Reviewer: >Can be highly effective.

Try to not drop the subject from sentences.

Author: I changed this (a39c3b5), and added the subject. I got the sense that always referring to `Browser Intrinsic` mechanisms is weird, so I will probably rename this to `No-Opt Mechanisms` in the future.

> Can be highly effective. The biggest advantage is that it transparently protects applications and users, and should not require extra efforts from developers.

Reviewer: ```suggestion
Can be highly effective. The biggest advantage is that it protects applications and users, and does not require extra efforts from developers.
```

Author: This got outdated when I applied the first suggestion. This suggestion was applied in b75bc7e.

{{< /hint >}}

## Opt-in Mechanisms

> These [Defense Mechanisms]({{< ref "opt-in/_index.md" >}}) allow applications to address a group of similar XS-Leaks as one. By deploying these protections an application can decide how requests to itself are made and make more intelligent decisions when something wrong is about to happen.

Reviewer: ```suggestion
These [Defense Mechanisms]({{< ref "opt-in/_index.md" >}}) allow applications to address classes of similar XS-Leaks at once. By deploying these protections an application can instruct browsers to behave in a more secure way or it can leverage some additional information to block potential attacks.
```
